lib/pbench/server/auth/auth.py

```diff
@@ -130,36 +190,58 @@ def verify_auth(auth_token):
         return None
 
     @staticmethod
-    def verify_third_party_token(auth_token: str, oidc_client: OpenIDClient) -> bool:
+    def verify_third_party_token(
+        auth_token: str,
+        oidc_client: OpenIDClient,
+        algorithms: Optional[list[str]] = ["HS256"],
+    ) -> "InternalUser":
         """
         Verify a token provided to the Pbench server which was obtained from a
         third party identity provider.
         Args:
             auth_token: Token to authenticate
             oidc_client: OIDC client to call to authenticate the token
+            algorithms: Optional token signature algorithm argument,
+                        defaults to HS256
         Returns:
             True if the verification succeeds else False
         """
-        identity_provider_pubkey = oidc_client.get_oidc_public_key(auth_token)
         try:
-            oidc_client.token_introspect_offline(
+            identity_provider_pubkey = oidc_client.get_oidc_public_key(auth_token)
+        except Exception:
+            Auth.logger.warning("Identity provider public key fetch failed")
+            identity_provider_pubkey = Auth().get_secret_key()
+        try:
+            audience = oidc_client.client_id if oidc_client else None
+            token_decode = oidc_client.token_introspect_offline(
                 token=auth_token,
                 key=identity_provider_pubkey,
-                audience=oidc_client.client_id,
+                audience=audience,
+                algorithms=algorithms,
                 options={
                     "verify_signature": True,
                     "verify_aud": True,
                     "verify_exp": True,
                 },
             )
-            return True
+            roles = []
+            if audience in token_decode.get("resource_access"):
+                roles = token_decode.get("resource_access").get(audience).get("roles")
```

Comment:
> The chained `get()` calls here are kind of silly. First, since they don't specify a default, if they fail they are going to return `None` which is going to result in an `AttributeError` which is much harder to diagnose than the `KeyError` which you would have gotten with a normal dictionary access.

Good point

Reply:
What we really want is the Javascript `?.` operator to cascade field references without worrying about a null; that'd be a useful addition to Python.